strings/base_implements.h

```diff
@@ -788,6 +788,12 @@ namespace winrt::impl
             return m_inner.try_as<Qi>();
         }
 
+        template <typename Qi>
+        auto try_as_with_reason() const noexcept
```

Comment:
Possibly.  I'll give it a try.

Reply:
Unfortunately we cannot.  The caller is the `winrt::impl::consume_` method which is outside that class and does not have access to protected members.  Here's an example break with that change in effect:

`v:\T\cppwinrt\_build\x64\Release\winrt\Windows.UI.Xaml.h(4408,86): error C2248: 'winrt::impl::root_implements_composing_outer<true>::try_as_with_reason': cannot access protected member declared in class 'winrt::impl::root_implements_composing_outer<true>' [v:\T\cppwinrt\test\old_tests\UnitTests\Tests.vcxproj]`

Reply:
We could pass it through a friend.
```cpp
struct root_implements_composing_outer
{
...
private:
template<typename Q, typename T>
friend auto impl::try_as_with_reason(T*);
};

and then
```cpp
namespace impl
{
template<typename Q, typename T>
auto impl::try_as_with_reason(T* t) { return t->template try_as_with_reason<Q>(); };

auto consume_blah()
{
 auto const& [castedResult, code] = impl:;try_as_with_reason<%>(static_cast<D const*>(this));
  ...
}

Reply:
It took me a few hours to implement that approach but I think I've got it.  cppwinrt itself (and the tests) compile.  I've got some other projects building to increase my confidence further.

Reply:
In this case we can use as

Reply:
Possibly.  I was hesitant to do so because it could over-match and do the wrong thing.  But now that I think about it further, as long as the `friend` method has a different name than just `as` it shouldn't match any existing code and that would be fine.

Reply:
Actually that's not quite right.  Whatever the new name is would have to exist for non-composing objects too.  So I either name it `.as()` and have some confusion with the other `as` methods, or I give it a different name and I'm back to adding a new variant of `as` method to every instance.  At that point I think I'm better off just keeping what is already here.